### PR TITLE
Remove some locks in ParallelExecutor

### PR DESCRIPTION
**Pull request by sneaxiy**
Remove the lock and record event in computation_op which does not have following all_reduce_op, etc.

Test condition: Transformer model. P40 GPU, 8 cards, batch_size = 4096, batch_num = 100
With lock: time cost 115s
Without lock: time cost 113s

### CODE REVIEW

File: paddle/fluid/framework/details/modify_op_lock_and_record_event_pass.h [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+class ModifyOpLockAndRecordEventPass : public ir::Pass {
+ protected:
+  std::unique_ptr<ir::Graph> ApplyImpl(
+      std::unique_ptr<ir::Graph> graph) const override;
+};
+
+}  // namespace details
+}  // namespace framework
+}  // namespace paddle
```
Comment (chengduoZH) [inline paddle/fluid/framework/details/modify_op_lock_and_record_event_pass.h:32]: `modify_op_lock_and_record_event_pass.h` is only used by `modify_op_lock_and_record_event_pass.cc`, so class definition and declaration can be wrote together.

File: paddle/fluid/framework/details/build_strategy.h [hunk 2]
```diff
@@ -73,6 +73,10 @@ struct BuildStrategy {
 
   bool fuse_broadcast_op_{false};
 
+#ifdef PADDLE_WITH_CUDA
```
Comment (reyoung) [inline paddle/fluid/framework/details/build_strategy.h]: Remove this macro, so we can give the same `API` between GPU and CPU.

File: paddle/fluid/framework/details/computation_op_handle.h [hunk 3]
```diff
@@ -28,14 +28,25 @@ namespace framework {
...
+  ComputationOpHandle(ir::Node *node, Scope *scope, platform::Place place,
+                      size_t scope_idx);
 
   std::string Name() const override;
 
   const Scope *GetScope() const { return scope_; }
 
   const platform::Place &GetPlace() const { return place_; }
 
+  size_t GetScopeIdx() const { return scope_idx_; }
```
Comment (reyoung) [inline paddle/fluid/framework/details/computation_op_handle.h]: What is the usage of `GetScopeIdx`?

File: paddle/fluid/framework/details/computation_op_handle.h [hunk 4]
```diff
@@ -28,14 +28,25 @@ namespace framework {
...
+
+  OperatorBase &GetOp() { return *op_; }
```
Comment (reyoung) [inline paddle/fluid/framework/details/computation_op_handle.h]: What is the usage of `GetOp`

File: paddle/fluid/framework/details/computation_op_handle.h [hunk 5]
```diff
@@ -45,6 +56,10 @@ struct ComputationOpHandle : public OpHandleBase {
   std::unique_ptr<OperatorBase> op_;
   Scope *scope_;
   platform::Place place_;
+  size_t scope_idx_;
+#ifdef PADDLE_WITH_CUDA
+  bool is_lock_and_record_event_free_{false};
```
Comment (reyoung) [inline paddle/fluid/framework/details/computation_op_handle.h:50]: Maybe we can just remove the macro. It is annoying and change the ABI between GPU/CPU

File: paddle/fluid/framework/details/op_handle_graph.h [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+#include <unordered_map>
+#include <unordered_set>
+#include <vector>
+#include "paddle/fluid/framework/details/op_handle_base.h"
+
+namespace paddle {
+namespace framework {
+namespace details {
+
+class OpHandleGraph {
```
Comment (reyoung) [inline paddle/fluid/framework/details/op_handle_graph.h]: A lot of methods in this class is not used by other modules. It seems that we could keep them in this class for further usage or we can keep this class clean until they are needed.

File: paddle/fluid/framework/details/computation_op_handle.cc [hunk 7]
```diff
@@ -20,18 +20,30 @@ namespace paddle {
...
 ComputationOpHandle::ComputationOpHandle(ir::Node *node, Scope *scope,
-                                         platform::Place place)
+                                         platform::Place place,
+                                         size_t scope_idx)
     : OpHandleBase(node),
       op_(framework::OpRegistry::CreateOp(*node->Op())),
       scope_(scope),
-      place_(place) {}
+      place_(place),
+      scope_idx_(scope_idx) {}
```
Comment (chengduoZH) [inline paddle/fluid/framework/details/computation_op_handle.cc]: It seems that `scope_idx_ ` is not used in `ComputationOpHandle `.

File: paddle/fluid/framework/details/op_handle_graph.h (hunk 6 above)
Comment (reyoung) [reply paddle/fluid/framework/details/op_handle_graph.h]: `OpHandleGraph` is a confusing name. It just another view of IRGraph which only contains op.

Perhaps `OpGraphView` is better. `View` means this class does not change any data, just another view of graph. `OpGraph` means this graph only contains operators.

File: paddle/fluid/platform/device_context.cc [hunk 8]
```diff
@@ -166,10 +166,7 @@ class CudnnHolder {
   void RunFunc(const std::function<void(void*)>& cudnn_func,
```
Comment (reyoung) [inline paddle/fluid/platform/device_context.cc]: It seems that this method is not used. Could we remove it?

File: paddle/fluid/platform/device_context.cc [hunk 9]
```diff
@@ -203,6 +212,24 @@ class CudnnHolder {
   std::mutex mtx_;
 };
 
+CudnnWorkspaceHandle::CudnnWorkspaceHandle(CudnnHolder* holder)
+    : holder_(holder) {}
+
+void CudnnWorkspaceHandle::RunFunc(const std::function<void(void*)>& cudnn_func,
```
Comment (reyoung) [inline paddle/fluid/platform/device_context.cc]: use 
```
template <typename Callback>
void RunFunc(Callback cudnn_func) {}
```

then compiler can inline this method call.

File: paddle/fluid/framework/details/op_handle_graph.h [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+  explicit OpHandleGraph(const std::vector<std::unique_ptr<OpHandleBase>> &ops);
+
+  size_t OpNumber() const;
+
+  std::unordered_set<OpHandleBase *> AllOps() const;
+
+  const std::unordered_set<OpHandleBase *> &PrecedingOps(
+      OpHandleBase *op) const;
+
+  const std::unordered_set<OpHandleBase *> &PendingOps(OpHandleBase *op) const;
```
Comment (chengduoZH) [inline paddle/fluid/framework/details/op_handle_graph.h]: If `PendingOps` is the only used function for `OpHandleGraph `, please remove other function which is not used in this PR.

File: paddle/fluid/framework/details/modify_op_lock_and_record_event_pass.h (hunk 1 above)
Comment (sneaxiy) [reply paddle/fluid/framework/details/modify_op_lock_and_record_event_pass.h:32]: Just for faster compiling.

File: paddle/fluid/framework/details/build_strategy.h (hunk 2 above)
Comment (sneaxiy) [reply paddle/fluid/framework/details/build_strategy.h]: Done. Thanks.

File: paddle/fluid/framework/details/computation_op_handle.h (hunk 3 above)
Comment (sneaxiy) [reply paddle/fluid/framework/details/computation_op_handle.h]: For future use. Have deleted.

File: paddle/fluid/framework/details/computation_op_handle.h (hunk 4 above)
Comment (sneaxiy) [reply paddle/fluid/framework/details/computation_op_handle.h]: For future use. Have deleted.

File: paddle/fluid/framework/details/computation_op_handle.h (hunk 5 above)
Comment (sneaxiy) [reply paddle/fluid/framework/details/computation_op_handle.h:50]: Done. Thanks.

File: paddle/fluid/framework/details/op_handle_graph.h (hunk 6 above)
Comment (sneaxiy) [reply paddle/fluid/framework/details/op_handle_graph.h]: For future use. Have deleted unused methods and change `OpHandleGraph` to be `OpGraphView`.

File: paddle/fluid/platform/device_context.cc (hunk 9 above)
Comment (sneaxiy) [reply paddle/fluid/platform/device_context.cc]: Done. Thanks.

File: paddle/fluid/framework/details/computation_op_handle.cc (hunk 7 above)
Comment (sneaxiy) [reply paddle/fluid/framework/details/computation_op_handle.cc]: For future use. Have deleted.

File: paddle/fluid/framework/details/op_handle_graph.h (hunk 10 above)
Comment (sneaxiy) [reply paddle/fluid/framework/details/op_handle_graph.h]: For future use. Have deleted unused methods.

**Review by reyoung — APPROVED** [review]
Excellent

**Pull request merged**